src/main/java/org/jabref/gui/externalfiles/ImportHandler.java

```diff
@@ -62,40 +71,96 @@ public ExternalFilesEntryLinker getLinker() {
+                                addResultToList(file, false, Localization.lang("Error reading PDF content: %0", pdfImporterResult.getErrorMessage()));
+                            }
+
+                            var xmpParserResult = contentImporter.importXMPContent(file);
+                            List<BibEntry> xmpEntriesInFile = xmpParserResult.getDatabase().getEntries();
+
+                            if (xmpParserResult.hasWarnings()) {
+                                addResultToList(file, false, Localization.lang("Error reading XMP content: %0", xmpParserResult.getErrorMessage()));
+                            }
+
+                            // First try xmp import, if empty try pdf import, otherwise create empty entry
+                            if (!xmpEntriesInFile.isEmpty()) {
+                                entriesToAdd = xmpEntriesInFile;
+                                addResultToList(file, true, Localization.lang("Importing using XMP data..."));
+                            } else if (!pdfEntriesInFile.isEmpty()) {
+                                    entriesToAdd = pdfEntriesInFile;
+                                    addResultToList(file, true, Localization.lang("Importing using extracted PDF data"));
+                            } else {
+                                    entriesToAdd = Collections.singletonList(createEmptyEntryWithLink(file));
+                                    addResultToList(file, false, Localization.lang("No metadata found. Creating empty entry with file link"));
+                            }
+                        } else if (FileUtil.isBibFile(file)) {
+                            var bibtexParserResult = contentImporter.importFromBibFile(file, fileUpdateMonitor);
+                            if (bibtexParserResult.hasWarnings()) {
+                                addResultToList(file, false, bibtexParserResult.getErrorMessage());
+                            }
+
+                            entriesToAdd = bibtexParserResult.getDatabaseContext().getEntries();
+                            addResultToList(file, false, Localization.lang("Importing bib entry"));
+                        } else {
+                            entriesToAdd = Collections.singletonList(createEmptyEntryWithLink(file));
+                            addResultToList(file, false, Localization.lang("No bibtex data found. Creating empty entry with file link"));
+                        }
+
+                    } catch (IOException ex) {
+                        LOGGER.error("Error importing", ex);
+                        addResultToList(file, false, Localization.lang("Error from import: %0", ex.getLocalizedMessage()));
+
+                        DefaultTaskExecutor.runInJavaFXThread(() -> updateMessage(Localization.lang("Error")));
                     }
+
+                    // We need to run the actual import on the FX Thread, otherwise we will get some deadlocks with the UIThreadList
```

Comment:
Really? I thought that should actually work. Where do you get deadlocks otherwise?

Reply:
I think the deadlock came from the database.insertEntries or so which fires the changed listeners